### PR TITLE
DM-18569: Precreate the results table in qserv

### PR DESCRIPTION
**Pull request by n8pease**

### CODE REVIEW

File: core/modules/rproc/InfileMerger.cc [hunk 1]
```diff
@@ -355,15 +348,72 @@ int InfileMerger::makeJobIdAttempt(int jobId, int attemptCount) {
...
+
+void InfileMerger::_addJobIdColumnToSchema(sql::Schema& schema) {
+    unsigned int attempt = 0;
+    auto columnItr = schema.columns.begin();
+    while (columnItr != schema.columns.end()) {
+        if (columnItr->name == _jobIdColName) {
+            _jobIdColName = JOB_ID_BASE_NAME + std::to_string(attempt++);
+            columnItr = schema.columns.begin(); // start over
+        }
+        ++columnItr;
```
Comment (andy-slac) [inline core/modules/rproc/InfileMerger.cc]: I think you want `else` before this line.

File: core/modules/rproc/InfileMerger.cc [hunk 2]
```diff
@@ -355,15 +348,72 @@ int InfileMerger::makeJobIdAttempt(int jobId, int attemptCount) {
...
+        return false;
+    }
+    LOGS(_log, LOG_LVL_DEBUG, "InfileMerger extracted schema: " << schema);
+
+    _addJobIdColumnToSchema(schema);
+
+    std::string createStmt = sql::formCreateTable(_mergeTable, schema);
+
+    // Specifying engine. There is some question about whether InnoDB or MyISAM is the better
+    // choice when multiple threads are writing to the result table.
```
Comment (andy-slac) [inline core/modules/rproc/InfileMerger.cc]: I know it was copied from old code but this comment does not make sense.

File: core/modules/rproc/InfileMerger.h [hunk 3]
```diff
@@ -219,14 +222,20 @@ class InfileMerger {
...
     std::mutex _sqlMutex; ///< Protection for SQL connection
-    bool _needCreateTable{true}; ///< Does the target table need creating?
     size_t _getResultTableSizeMB(); ///< Return the size of the result table in MB.
-    /// Alter the jobId column name in hopes that it will be unique.
-    void _alterJobIdColName() {
-        _jobIdColName = "jobId" + std::to_string(_jobIdColNameAdj++);
-    }
+
+    /**
+     * @brief Put a "jobId" column first in the provided schema.
```
[the remark this answers is not in the document]
Comment (n8pease) [reply core/modules/rproc/InfileMerger.h]: yes

File: core/modules/qana/QueryPlugin.h [hunk 4]
```diff
@@ -86,15 +86,17 @@ class QueryPlugin {
 class QueryPlugin::Plan {
 public:
     Plan(query::SelectStmt& stmtOriginal_, SelectStmtPtrVector& stmtParallel_,
-         query::SelectStmt& stmtMerge_, bool hasMerge_)
+         std::shared_ptr<query::SelectStmt>& stmtLocalWorker_, query::SelectStmt& stmtMerge_, bool hasMerge_)
```
[the remark this answers is not in the document]
Comment (n8pease) [reply core/modules/qana/QueryPlugin.h]: in this case no, it's a reference to a pointer that does get changed. Plan is an interface into QuerySession that allows plugins to change members of QuerySession. It's kind of wack. But it also makes sense in a wack kind of way. Anyway, I decided not to change it for now at least.

[the remark this answers is not in the document]
Comment (n8pease) [reply core/modules/qana/QueryPlugin.h]: I'll add some commentage tho.

File: core/modules/rproc/InfileMerger.h (hunk 3 above)
[the remark this answers is not in the document]
Comment (andy-slac) [reply core/modules/rproc/InfileMerger.h]: Hmm, it is a little troubling that we do not enforce this sort of agreement in any way. Basically we build some schema locally and trust workers that they return their data according to this schema (which they do not even see).

[the remark this answers is not in the document]
Comment (n8pease) [reply core/modules/rproc/InfileMerger.h]: The workers do provide schema with results, we could probably compare. Or, send “expected” schema to workers and allow them to confirm before sending results. Or can you think f another way we should be testing it?

[the remark this answers is not in the document]
Comment (n8pease) [reply core/modules/rproc/InfileMerger.h]: After discussing this with the team on Slack, I created DM-18898 to track work for this issue.

**Pull request merged**